R/split_data_by_task.R

```diff
@@ -28,16 +30,43 @@
 #'     oracle_value = observation
 #'   )
 #' valid_tbl <- validate_input_data(forecast_data, target_data)
-#' split_data_by_task(valid_tbl)
+#' split_data_by_task(valid_tbl, weighted = FALSE, training_window_length = 0)
+#' split_data_by_task(valid_tbl, weighted = TRUE, training_window_length = 1)
 #' }
-split_data_by_task <- function(valid_tbl) {
-  # Get columns to use for splitting data by task
-  split_cols <- c("horizon", "location", "target_end_date")
+split_data_by_task <- function(valid_tbl, weighted, training_window_length) {
+  # Use of a trained ensemble or not.
+  training_status <- ifelse(weighted, "trained", "untrained")
 
-  # List of data sets split by task
-  datasets_by_task <- valid_tbl |>
-    dplyr::group_by(across(all_of(split_cols))) |>
-    dplyr::group_split()
+  # Split data and make a list of data sets
+  if (training_status == "untrained") {
+    # Get columns to use for splitting data by task
+    split_cols <- c("horizon", "location", "target_end_date")
+    # List of data split by task
+    list_datasets <- valid_tbl |>
+      dplyr::group_by(across(all_of(split_cols))) |>
+      dplyr::group_split()
+  } else {
+    # List of reference dates
+    all_ref_dates <- sort(unique(valid_tbl$reference_date))
+    # Length of reference dates
+    len_ref_dates <- length(all_ref_dates)
+    if (len_ref_dates <= training_window_length) {
+      stop(
+        "The number of reference_date must greater than the training window
+        length."
+      )
+    } else {
+      # Create a list of data split by reference date including the training
+      list_datasets <- list()
+      for (i in 1:(len_ref_dates - training_window_length)) {
+        list_datasets[[i]] <- valid_tbl |>
+          dplyr::filter(
+            .data$reference_date <= all_ref_dates[i + training_window_length],
+            .data$reference_date >= all_ref_dates[i]
+          )
+      }
```

Comment:
I remember learning something a long time ago about how this kind of recursive assignment/creation of objects can be an inefficient use of memory. It may be fine here, but I could imagine there might be long lists of a lot of data, so it might be the kind of thing where it could matter. This is a place where @matthewcornell might have suggestions of whether there might be a better way of structuring the code.

Reply:
Good question, Nick. I'm not familiar with R's efficiency of loops like the for() vs. dplyer and other dataframe operations. I believe the latter are faster, but I'm not 100% sure.

Reply:
You can use a for loop syntax if you pre-allocate the list. Mapping or dplyr operations are not necessarily going to be better or faster for this operation. 


```suggestion
      n_dates <- len_ref_dates - training_window_length
      list_datasets <- vector(mode = "list", length = n_dates)
      for (i in seq_len(n_dates)) {
        list_datasets[[i]] <- valid_tbl |>
          dplyr::filter(
            .data$reference_date <= all_ref_dates[i + training_window_length],
            .data$reference_date >= all_ref_dates[i]
          )
      }
```

Reply:
Thanks @zkamvar . Just to clarify, one "could" do this either way (the way Minsu has it currently coded, or your suggested way) but you recommend the pre-allocation for more efficient memory handling?

Reply:
I would think a parallelized operation should be more efficient.

But if I'm understanding the workflow here, the final list has a lot of repeat data, correct? Then something is done to each of the data frames in the list?

Maybe you could make the inside of that for() loop a function which takes an index and a training_window_length and then apply that in parallel, or create the data frames at a time and then operate on each data frame as needed, then discard and on to the next, so you're holding so much repeat data in memory.

Reply:
> but you recommend the pre-allocation for more efficient memory handling?

Yes. I will also note that my proposal also fixes the `1:n` form of generating a sequence, [which can have problems if `n` is zero](https://lintr.r-lib.org/reference/seq_linter.html).

Reply:
Hi everyone, thank you all so much for this discussion! I truly appreciate each of you. 
@zkamvar, Thanks a lot for your code. I replaced my code with yours.